midx.c

```diff
@@ -832,15 +836,43 @@ static int write_midx_internal(const char *object_dir, struct multi_pack_index *
 
 	for_each_file_in_pack_dir(object_dir, add_pack_to_midx, &packs);
 
-	if (packs.m && packs.nr == packs.m->num_packs)
+	if (packs.m && packs.nr == packs.m->num_packs && !packs_to_drop)
 		goto cleanup;
 
+	sort_packs_by_name(packs.names, packs.nr, packs.perm);
+
+	if (packs_to_drop && packs_to_drop->nr) {
+		int drop_index = 0;
+		int j;
+
+		/* truncate the list of packs */
```

Comment:
nit: the word "truncate" is misleading here.  to me that says to lop of the end.  but you're collapsing the newly created holes.